### ISSUE TITLE
TypeScript Local Executors

### ISSUE THREAD

**Issue by flippidippi** [nrwl/nx#6926]
## Description

Similar to generators `index.ts`, local project executors should work without needing to run `tsc` on TypeScript `impl.ts`. 

## Motivation

Better users experience as users wouldn't need to run `tsc` for local custom executors or commit the `imp.js` to source code.

## Suggested Implementation

When executors run across a `impl.ts` and no `impl.js`, similar to generators, compile the TypeScript first then run.

**Comment by Pouja** [comment]
We are currently working around this issue by having the `impl.js` look like:

```js
require('ts-node').register({
  project: 'tsconfig.base.json',
});

module.exports = require('./your-executor.ts');
```

Where `your-executor.ts` exports an executor that implements `Executor` from `@nrwl/tao/src/shared/workspace`

**Comment by edmundo096** [comment]
This was asked around 2 years ago (before Nx separated a bit more from Angular CLI/devkit I think and started to call them more properly "executors" instead of Angular CLI's "builders").

Linking the previous issue #2211 where more discussion was made back then (was automatically closed, hope this newer one doesn't).

Back then we didn't even had the suggestion to manually compile them with "tsc", and the issue was more because Angular CLI/devkit didn't supported that, but this may have changed as how Nx have matured and seem to have separated more from Angular's CLI/devkit.

Hope this can be reviewed again now that it was asked once more time.

**Comment by jasongerbes** [comment]
@FrozenPandaz do you have any input to share on this? I feel there should be better support for workspace/local executors. 

The [documented solution](https://nx.dev/executors/creating-custom-builders#compiling-and-running-your-executor) of manually running `tsc` seems problematic as you would need to manually re-run `tsc` whenever the source code changes.

**Comment by AgentEnder** [comment]
This will be fixed by #9116.
